File: Algorithms/[1, 1k)/962/segment_tree.swift
class Solution {
    
    var tree = Array(repeating: -1, count: 1 << 17)
    
    func update(index: Int, offset: Int) {
        if index <= 0 || offset <= tree[index] { return }
        
        tree[index] = offset
        update(index: index >> 1, offset: offset)
    }
    
    func find(index: Int=1, range: Range<Int>=0..<(1<<16), target: Int) -> Int {
        if target >= range.upperBound { return -1 }
        if target <= range.lowerBound || tree[index] == -1 { return tree[index] }
        
        let mid = (range.lowerBound + range.upperBound) >> 1
        return max(
            find(index: index<<1, range: range[..<mid], target: target),
            find(index: index<<1|1, range: range[mid...], target: target)
        )
    }
    
    func maxWidthRamp(_ nums: [Int]) -> Int {
        for i in 0..<tree.count {
            tree[i] = -1
        }
        
        var opt = 0
        for (i, v) in nums.enumerated().reversed() {
            let j = max(i, find(target: v))
            opt = max(opt, j - i)
            update(index: (1 << 16) + v, offset: i)
        }
        return opt
    }
}

File: Algorithms/[1k, 2k)/1425/segment_tree.go
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

var _tree = make([]int, 4e5)

func findMax(i, l, r, wL, wR int) int {
	if wL <= l && r <= wR {
		return _tree[i]
	}

	m := (l + r) / 2
	if wR <= m {
		return findMax(i*2+1, l, m, wL, wR)
	} else if m < wL {
		return findMax(i*2+2, m+1, r, wL, wR)
	}
	return max(findMax(i*2+1, l, m, wL, m), findMax(i*2+2, m+1, r, m+1, wR))
}

func update(i, l, r, t, v int) {
	for {
		_tree[i] = max(_tree[i], v)
		if l == r {
			break
		}

		m := (l + r) / 2
		if t <= m {
			i = i*2 + 1
			r = m
		} else {
			i = i*2 + 2
			l = m + 1
		}
	}
}

func constrainedSubsetSum(nums []int, k int) int {
	for i := 0; i < len(nums)<<2; i++ {
		_tree[i] = -1e5
	}

	update(0, 0, len(nums)-1, 0, nums[0])
	for i := 1; i < len(nums); i++ {
		v := findMax(0, 0, len(nums)-1, max(0, i-k), i-1)
		v = max(v+nums[i], nums[i])
		update(0, 0, len(nums)-1, i, v)
	}
	return _tree[0]
}